crate_universe/extension.bzl

```diff
@@ -0,0 +1,224 @@
+    repo_name = "crates"
+    if cfg.suffix:
+        tag_path = mod_path.get_child(cfg.suffix)
+        repo_name += "_" + cfg.suffix
+
+    cargo_lockfile = module_ctx.path(cfg.cargo_lockfile)
+
+    rendering_config = json.decode(render_config(
+        regen_command = "Run 'cargo update [--workspace]'",
+    ))
+    config_file = tag_path.get_child("config.json")
+    module_ctx.file(
+        config_file,
+        executable = False,
+        content = generate_config_file(
+            module_ctx,
+            mode = "remote",
+            annotations = annotations,
+            generate_build_scripts = cfg.generate_build_scripts,
+            supported_platform_triples = cfg.supported_platform_triples,
+            generate_target_compatible_with = True,
+            repository_name = repo_name,
+            output_pkg = repo_name,
+            workspace_name = repo_name,
+            generate_binaries = cfg.generate_binaries,
+            rendering_config = rendering_config,
+        ),
+    )
+
+    manifests = {module_ctx.path(m): m for m in cfg.manifests}
+    splicing_manifest = tag_path.get_child("splicing_manifest.json")
+    module_ctx.file(
+        splicing_manifest,
+        executable = True,
+        content = generate_splicing_manifest(
+            packages = {},
+            splicing_config = "",
+            cargo_config = cfg.cargo_config,
+            manifests = {str(k): str(v) for k, v in manifests.items()},
+            manifest_to_path = module_ctx.path,
+        ),
+    )
+
+    splicing_output_dir = tag_path.get_child("splicing-output")
+    cargo_bazel([
+        "splice",
+        "--output-dir",
+        splicing_output_dir,
+        "--config",
+        config_file,
+        "--splicing-manifest",
+        splicing_manifest,
+        "--cargo-lockfile",
+        cargo_lockfile,
+    ])
+
+    # Create a lockfile, since we need to parse it to generate spoke
+    # repos.
+    lockfile_path = tag_path.get_child("lockfile.json")
+    module_ctx.file(lockfile_path, "")
+
+    # cargo-bazel generate takes the lockfile as input, but also writes
+    # to the lockfile. This means that even though nothing changes, the
+    # modified timestamp of the file is updated. Since the lock file is
+    # an input to the rule, this would invalidate the repo rule,
+    # requiring it to be rerun on every invocation.
+    # To solve this, we allow it to touch a copy of the lock file,
+    # rather than the original.
+    cargo_lockfile_copy = tag_path.get_child("copy/Cargo.lock")
+    module_ctx.file(
+        cargo_lockfile_copy,
+        module_ctx.read(cargo_lockfile),
+    )
+    cargo_bazel([
+        "generate",
+        "--cargo-lockfile",
+        cargo_lockfile_copy,
+        "--config",
+        config_file,
+        "--splicing-manifest",
+        splicing_manifest,
+        "--repository-dir",
+        tag_path,
+        "--metadata",
+        splicing_output_dir.get_child("metadata.json"),
+        "--repin",
```

Comment:
This looks like it unconditionally repins? This is quite different from how the non-bzlmod logic works (https://github.com/bazelbuild/rules_rust/blob/86379418f04b0144cbf3959641ac89797a7c9a10/crate_universe/private/crates_repository.bzl#L44-L86)

What's the story with the difference here? Is this an intermediate step to making pinning conditional, or is there some reason bzlmod needs to do pinning differently?

Reply:
I think the easiest way to explain this is the different workflow.

If I wanted to update `anyhow` from `1.0.77` to `1.0.78`, the workflow here is:
1. Update the entry in `Cargo.toml` from `1.0.77` to `1.0.78`
2. Run `cargo update --workspace` to update `Cargo.lock`

Since everything is transient and automatically generated, you can then just build and it works. I'm not personally a big fan of the whole `CARGO_BAZEL_REPIN` method, but if people do care about it, I think it's fine to add it in a follow-up PR, but it definitely isn't needed for the MVP.

Note that if you try and build between step 1 and 2 you'll get an error saying that we need to update the lockfile, but that `--locked` was passed, so it'll refuse to compile.